### PR TITLE
Re-work yast2 control center and fix timeout issue

### PR DESCRIPTION
**Pull request by Zaoliang**
Please check reference test at: http://e13.suse.de/tests/2111

### CODE REVIEW

File: tests/yast2_gui/yast2_control_center.pm [hunk 1]
```diff
@@ -8,7 +8,7 @@
 # notice and this notice are preserved.  This file is offered as-is,
 # without any warranty.
 
-# Summary: Extend YaST2 UI test yast2-control-center
+# Summary: Make YaST2 UI test yast2-control-center more reliable and avoid timeout error
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: please don't describe the change within your commit here but what the test module does in the current state

File: tests/yast2_gui/yast2_control_center.pm [hunk 2]
```diff
@@ -33,85 +33,110 @@ sub run() {
         send_key 'backspace';
     }
 
+
+    #	increase timeout value to 60 seconds for assert_screen and check_screen where is required
     #	start Add-on Products
     assert_and_click "yast2_control-center_add-on";
-    assert_screen "yast2_control-center_add-on_installed";
-    send_key "alt-c";
+    if (assert_screen "yast2_control-center_add-on_installed") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: that does not make sense. assert_screen never returns with false, it dies the test

File: tests/yast2_gui/yast2_control_center.pm [hunk 3]
```diff
@@ -33,85 +33,110 @@ sub run() {
         send_key 'backspace';
     }
 
+
+    #	increase timeout value to 60 seconds for assert_screen and check_screen where is required
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: stating the obvious, delete the comment

File: tests/yast2_gui/yast2_control_center.pm (hunk 1 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: okay, change it back now

File: tests/yast2_gui/yast2_control_center.pm (hunk 3 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: done

File: tests/yast2_gui/yast2_control_center.pm (hunk 2 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: hm... then I need to check this overall

Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: checking now all if asset_screen in test code and will provide changes soon.

File: tests/yast2_gui/yast2_control_center.pm (hunk 1 above)
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: no, that does not help. "Extend UI test" is also not a good summary of what the test module does

Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: How about following:
Summary: YaST2 UI test yast2-control-center provides sanity checks for YaST modules

I've extended initial version, so that should be okay regarding to the changes.

Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: perfect

**Comment by Zaoliang** [comment]
Please see my new reference test at: http://e13.suse.de/tests/2193
Thanks!

File: tests/yast2_gui/yast2_control_center.pm [hunk 4]
```diff
@@ -47,71 +48,83 @@ sub run() {
...
     #	start Software Repositiories
     assert_and_click "yast2_control-center_software-repositories";
     assert_screen "yast2_control-center_configured-software-repositories";
-    assert_and_click "yast2_control-center_configured-software-repositories_ok";
+    send_key "alt-o";
     assert_screen 'yast2-control-center-ui';
 
     #	start Printer
     assert_and_click "yast2_control-center_printer";
+    if (check_screen "yast2_control-center_printer_running-cups-daemon") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: that can be replaced by

```perl
assert_screen [qw(yast2_control-center_printer_running-cups-daemon yast2_control-center_printer_configurations)];
if (match_has_tag 'yast2_control-center_printer_running-cups-daemon') {
    …
    assert_screen 'yast2_control-center_printer_configurations';
}
```

File: tests/yast2_gui/yast2_control_center.pm [hunk 5]
```diff
@@ -47,71 +48,83 @@ sub run() {
...
-    assert_screen "yast2_control-center_bootloader_configuration";
-    assert_and_click "yast2_control-center_bootloader_ok";
-    assert_screen 'yast2-control-center-ui';
+    assert_screen "yast2_control-center_bootloader_ok";
+    send_key "alt-o";
+    assert_screen 'yast2-control-center-ui', 60;
 
     #	start Date and Time
     assert_and_click "yast2_control-center_date-and-time";
+    if (check_screen "yast2_control-center_data-and-time_ntp.conf_changed") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 6]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
 
     #   start Authentication Server
     #	need to scroll down to get other modules
     send_key "down";
     assert_and_click "yast2_control-center_authentication-server";
-    assert_screen "yast2_control-center-authentication-server_install";
-    assert_and_click "yast2_control-center-authentication-server_install-cancel";
-    assert_and_click "yast2_control-center-authentication-server_install-error";
-    assert_screen 'yast2-control-center-ui';
+    if (check_screen "yast2_control-center-authentication-server_install") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: can you create a needle for both cases "if" and "else", check on both with `assert_screen` and use `match_has_tag`?

File: tests/yast2_gui/yast2_control_center.pm [hunk 7]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+        send_key "alt-o";
+        assert_screen 'yast2-control-center-ui';
+    }
 
     #   start DHCP Server
     assert_and_click "yast2_control-center_dhcp-server";
-    assert_and_click "yast2_control-center-dhcp-server-install_cancel";
-    assert_and_click "yast2_control-center-dhcp-server-install-cancel_again";
-    assert_screen 'yast2-control-center-ui';
+    if (check_screen "yast2_control-center-dhcp-server-install_cancel") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 8]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+        check_screen "yast2_control-center-dhcp-server-install-cancel_again";
+        send_key "alt-c";
+    }
+    assert_screen 'yast2-control-center-ui', 60;
 
     #   start DNS Server
     assert_and_click "yast2_control-center_dns-server";
-    assert_and_click "yast2_control-center-dns-server-install_cancel";
-    assert_and_click "yast2_control-center-dns-server-install_abort";
+    if (check_screen "yast2_control-center-dns-server-install_cancel", 60) {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 9]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+        send_key "alt-c";
+    }
     assert_screen 'yast2-control-center-ui';
 
     #   start FTP Server
     assert_and_click "yast2_control-center_ftp-server";
-    assert_screen "yast2_control-center_ftp-start-up";
-    assert_and_click "yast2_control-center_ftp-server_cancel";
+    assert_screen "yast2_control-center_ftp-start-up", 60;
+    if (check_screen "yast2_control-center_ftp-server_cancel") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 10]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+    assert_screen 'yast2-control-center-ui', 60;
 
     #   start NTP Configuration
     assert_and_click "yast2_control-center_ntp-configuration";
-    assert_and_click "yast2_control-center_ntp-configuration_cancel";
+    assert_screen "yast2_control-center_ntp-configuration_cancel", 60;
+    send_key "alt-c";
+    assert_screen "yast2_control-center_ntp-configuration_error";
+    send_key "alt-o";
+    if (check_screen "yast2_control-center_ntp-configuration_advanced") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 11]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+    if (check_screen "yast2_control-center_samba-server-installation_abort", 60) {
+        send_key "alt-c";
+        if (check_screen "yast2_control-center_samba-server_rename-share") {
+            send_key "alt-r";
+        }
+    }
+    else {
+        assert_screen "yast2_control-center_samba-server_samba-configuration";
+        send_key "alt-c";
+    }
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 12]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
     assert_screen 'yast2-control-center-ui';
 
     #   start Squid Server
     assert_and_click "yast2_control-center_squid-server-configuration";
-    assert_and_click "yast2_control-center_squid-server-start-up_cancel";
-    assert_and_click "yast2_control-center_squid-server-installation_cancel";
+    if (check_screen "yast2_control-center_squid-server-start-up_cancel") {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 13]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+        send_key "alt-c";
+        check_screen "yast2_control-center_squid-server-installation_cancel";
+        send_key "alt-c";
+    }
     assert_screen 'yast2-control-center-ui';
 
     #   start TFTP Server
     assert_and_click "yast2_control-center_tftp-server-configuration";
-    assert_and_click "yast2_control-center_tftp-server-configuration_cancel";
+    if (check_screen "yast2_control-center_tftp-server-configuration_cancel", 60) {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 14]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
+    assert_screen "yast2_control-center_firewall_configuration", 60;
+    send_key "alt-c";
     assert_screen 'yast2-control-center-ui';
 
     #   start Linux Audit Framework LAF
     assert_and_click "yast2_control-center_laf";
-    assert_and_click "yast2_control-center_laf_cancel";
-    assert_and_click "yast2_control-center_laf_abort";
-    assert_and_click "yast2_control-center_laf_abort_really";
+    if (check_screen "yast2_control-center_laf_cancel", 60) {
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: see above

File: tests/yast2_gui/yast2_control_center.pm [hunk 15]
```diff
@@ -47,71 +48,83 @@ sub run() {
...
     #	start Printer
     assert_and_click "yast2_control-center_printer";
+    if (check_screen "yast2_control-center_printer_running-cups-daemon") {
+        send_key "alt-n";
+        check_screen "yast2_control-center_printer_running-cups-daemon_error";
+        send_key "alt-o";
+        check_screen "yast2_control-center_detect-printer-queues_error";
+        send_key "alt-o";
+        check_screen "yast2_control-center_show-printer-queues_error";
+        send_key "alt-o";
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: why should all these errors be acceptable?

File: tests/yast2_gui/yast2_control_center.pm [hunk 16]
```diff
@@ -123,220 +136,292 @@ sub run() {
...
     assert_and_click "yast2_control-center_http";
-    assert_and_click "yast2_control-center_http_finish";
+    assert_screen "yast2_control-center_http_finish";
+    send_key "alt-f";
     assert_screen 'yast2-control-center-ui';
 
     #   start iSCSI initiator
     assert_and_click "yast2_control-center_iscsi-initiator";
-    assert_and_click "yast2_control-center_iscsi-intiator_cancel";
+    assert_screen "yast2_control-center_iscsi-intiator_cancel", 60;
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: typo s/intiator/initiator/

File: tests/yast2_gui/yast2_control_center.pm (hunk 4 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: yes, I can change this for sure. Testing this right now.

File: tests/yast2_gui/yast2_control_center.pm (hunk 15 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: because I don't want to let restart cups, Anyway I will check this section and can change this a little bit and see what is better to allow to restart cups locally if delay is not too long.

File: tests/yast2_gui/yast2_control_center.pm (hunk 5 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: typo got changed already, needle tag got changed as well.

File: tests/yast2_gui/yast2_control_center.pm (hunk 16 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: corrected now

File: tests/yast2_gui/yast2_control_center.pm (hunk 6 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: okay, will create a new needle here. I can even check the results of package installation. It should like following:
--
assert_and_click "yast2_control-center_authentication-server";
assert_screen [qw[yast2_control-center-authentication-server_install yast2_control-center-authentication-server_configuration )];
if (match_has_tag ('yast2_control-center-authentication-server_install')) {
send_key "alt-i";
 }
else {  
send_key "alt-c"; 
}
assert_screen 'yast2-control-center-ui', 60;

File: tests/yast2_gui/yast2_control_center.pm (hunk 15 above)
Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: well, after I checked common behavior of cups service, that is normally not enabled. So it will restart  cups locally and it means it waits 30 seconds at least before to go next check...
It looks like following:
--
assert_screen [qw(yast2_control-center_printer_running-cups-daemon yast2_control-center_printer_configurations)];
 if (match_has_tag "yast2_control-center_printer_running-cups-daemon") {
 send_key "alt-y";
}
assert_screen "yast2_control-center_printer_configurations";
send_key "alt-o";
assert_screen 'yast2-control-center-ui', 60;
--

Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: think about your use of markup here :-D

Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: :) 
Still have a problem here. need to use check_screen here because cups-daemon dialogue pops up after printer_configuration. It should be like this:

```
assert_screen "yast_control-center_printer_configurations";
    send_key "alt-o";
    if (check_screen "yast2_control-center_printer_running-cups-daemon") { 
        send_key "alt-y";
    }
assert_screen 'yast2-control-center-ui', 60;
```

Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: As we talked this morning, I've updated this part and add other small changes, please check my reference test at: http://e13.suse.de/tests/2233
Thanks!

**Comment by okurz** [comment]
looks better. I would like to see the other `check_screen` replaced as well with multi-tag `assert_screen`

**Comment by Zaoliang** [comment]
Please check the changes, thanks!

File: tests/yast2_gui/yast2_control_center.pm [hunk 17]
```diff
@@ -33,6 +33,7 @@ sub run() {
         send_key 'backspace';
     }
 
+
```
Comment (okurz) [inline tests/yast2_gui/yast2_control_center.pm]: unrelated change, please remove

Comment (Zaoliang) [inline tests/yast2_gui/yast2_control_center.pm]: okay, empty line got removed

Please check reference test: http://e13.suse.de/tests/2312 and /tests/2314
Since there is a local server issue, it is hard to get all tests down without failure. But I am sure that that all tests after Windows Domain Membership are okay..

**Comment by okurz** [comment]
I would merge it but @coolo @sysrich please take a second (and/or third) look

**Comment by Zaoliang** [comment]
cool, I'm working right now on yast2_control_center for TW. That will help me a lot then when it runs on openQA soon.

**Comment by okurz** [comment]
test module is broken for a long time anyway so let's see how it behaves in production. Please make sure to closely monitor execution on both osd and o3 and fix
